SQL: Implement IN(value1, value2, ...) expression.

matriv opened the pull request:
Implement the functionality to translate the
`field IN (value1, value2,...) expressions to proper Lucene queries
or painless script depending on the use case.

The `IN` expression can be used in SELECT, WHERE and HAVING clauses.

Closes: #32955

elasticmachine commented:
Pinging @elastic/es-search-aggs

matriv commented:
Currently there is no check for null handling. It will be done in a next step once #34573 is merged.

There are no tests for negation when there is a translation to painless as this: #34558 must be fixed first

astefan commented:
I tested this a bit and a combination of a function and `IN` values of a different type (`select * from test_emp where lcase(first_name) in (1,2)`) would be incorrect, but the error message presented to the user is rather verbose and exposing the full stack trace of the error.

astefan commented:
@matriv this one fails and I think it shouldn't:

```
sql> select * from test_emp where year(birth_date) in (1990);
Server error [Server sent bad type [search_phase_execution_exception]. Original type was [all shards failed]. [Failed to execute phase [query], all shards failed; shardFailures {[1xf7J15UQ6-6Eagq0pzBBg][test_emp][0]: RemoteTransportException[[node-0][127.0.0.1:9300][indices:data/read/search[phase/query]]]; nested: QueryShardException[failed to create query: {
  "script" : {
    "script" : {
      "source" : "InternalSqlScriptUtils.dateTimeChrono(doc[params.v0].value.millis, params.v1, params.v2)==1990",
      "lang" : "painless",
      "params" : {
        "v0" : "birth_date",
        "v1" : "UTC",
        "v2" : "YEAR"
      }
    },
    "boost" : 1.0
  }
}]; nested: ScriptException[compile error]; nested: ClassCastException[Cannot cast from [int] to [java.lang.Object].]; }
        at org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseFailure(AbstractSearchAsyncAction.java:293)
        at org.elasticsearch.action.search.AbstractSearchAsyncAction.executeNextPhase(AbstractSearchAsyncAction.java:133)
        at org.elasticsearch.action.search.AbstractSearchAsyncAction.onPhaseDone(AbstractSearchAsyncAction.java:254)
```

costin commented:
The above comparison is fixed in #34573 (the underlying null-safe equality does proper widening when comparing `Number`s).

astefan commented:
```
sql> select * from test_emp where 1<2 in (false);
Bad request [Found 1 problem(s)
line 1:8: Unexecutable item]
```

Strange that the one above fails, but the next one doesn't: `select * from test_emp where 1<2 in (true)`.

matriv commented:
> select * from test_emp where 1<2 in (false);

This is a general problem, not related to in. Opened issue #34613.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/operator/comparison/InPipe.java:
```diff
@@ -0,0 +1,104 @@
+    private List<Pipe> right;
+
+    public InPipe(Location location, Expression expression, Pipe left, List<Pipe> right) {
+        super(location, expression, Stream.concat(Stream.of(left), right.stream()).collect(Collectors.toList()));
```
I'd opt for passing all the pipes in a list. Avoid creating the stream inside the constructor since the calling class already does the unwinding.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/operator/comparison/InPipe.java:
```diff
@@ -0,0 +1,104 @@
+        for (Pipe p : right) {
+            newRight.add(p.resolveAttributes(resolver));
+        }
+        return replaceChildren(Stream.concat(Stream.of(newLeft), newRight.stream()).collect(Collectors.toList()));
```
Again, if there's only one list no need to separate, concatenate things.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/operator/comparison/InPipe.java:
```diff
@@ -0,0 +1,104 @@
+
+    @Override
+    public boolean resolved() {
+        return left().resolved() && right().stream().allMatch(Pipe::resolved);
```
`Resolvables.resolved(list)`.

matriv replied:
hm cannot use that. `Pipe` is not `Resolvable`.

matriv replied:
Changing `Pipe` to implement `Resolvable`

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/operator/comparison/InPipe.java:
```diff
@@ -0,0 +1,104 @@
+
+public class InPipe extends Pipe {
+
+    private Pipe left;
```
I think left and right are a bit confusing - why not use value and list?

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryTranslator.java:
```diff
@@ -572,6 +579,55 @@ private static Query translateQuery(BinaryComparison bc) {
+            // if the code gets here it's a bug
+            //
+            else {
+                throw new UnsupportedOperationException("No idea how to translate " + in.value());
```
Use `SqlIllegalArgumentException` instead.

matriv replied:
Copy pasted from here: https://github.com/elastic/elasticsearch/pull/34581/files/ab7c1502d1b6341b2b3cc3be7eda72b00e606ada#diff-5d7529d13a2ef47d436ea2aa577e0c52R543 :-( Fixing both!

costin commented on x-pack/plugin/sql/src/test/java/org/elasticsearch/xpack/sql/planner/QueryTranslatorTests.java:
```diff
@@ -20,30 +20,40 @@
-    private Analyzer analyzer;
-    
-    public QueryTranslatorTests() {
+public class QueryTranslatorTests extends AbstractBuilderTestCase {
```
Is `AbstractBuilderTestCase` used anywhere (maybe I'm missing it)?

matriv replied:
I used it for the shardContext here: https://github.com/elastic/elasticsearch/pull/34581/files/ab7c1502d1b6341b2b3cc3be7eda72b00e606ada#diff-aef2b0ce456b8fdd5cc09d6cfd55f0c2R173.
I tried to manually mock it but ended up with ugly code. The AbstractBuilderTestCase is parent class for many tests in `.search.aggregations` and `.index.query` packages.

costin commented:
By the way, this should go in 6.5 as well.

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/analysis/analyzer/Verifier.java:
```diff
@@ -488,4 +503,31 @@ private static void checkNestedUsedInGroupByOrHaving(LogicalPlan p, Set<Failure>
\ No newline at end of file
+
+    private static void validateInExpression(Expression e, Set<Failure> localFailures) {
+        if (e instanceof In) {
```
Nice! Thx for the suggestion.

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/In.java:
```diff
@@ -54,29 +68,96 @@ public Expression value() {
+
+    @Override
+    protected Pipe makePipe() {
+        return new InPipe(location(), this, Expressions.pipe(value()), list.stream().map(Expressions::pipe).collect(Collectors.toList()));
```
No reason, It's the `InProcessor` that can do the separation. No reason to keep the value separate here. Merging into 1 list.

matriv commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/In.java:
```diff
@@ -54,29 +68,96 @@ public Expression value() {
     @Override
     public ScriptTemplate asScript() {
-        throw new SqlIllegalArgumentException("not implemented yet");
+        StringJoiner sj = new StringJoiner(" || ");
```
I already have this: #34582 to take care of null handling and add tests.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/planner/QueryTranslator.java:
```diff
@@ -144,7 +150,7 @@
 
     static QueryTranslation toQuery(Expression e, boolean onAggs) {
-        QueryTranslation translation = null;
+        QueryTranslation translation;
```
Looks like a formatting issue - I'd keep the explicit initialization (to keep introspection tools at bay that yes, `null` is expected).

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/querydsl/query/TermsQuery.java:
```diff
@@ -0,0 +1,58 @@
+    public TermsQuery(Location location, String term, List<Expression> values) {
+        super(location);
+        this.term = term;
+        this.values = values.stream().map(Expression::fold).collect(Collectors.toList());
```
the collector could be a set (as oppose to a list) to remove duplicates.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/operator/comparison/Comparisons.java:
```diff
@@ -5,34 +5,42 @@
  * Comparison utilities.
  */
-abstract class Comparisons {
+public final class Comparisons {
```
Minor nitpick, I tend to explicitly use `valueOf`/`xxxValue` to clarify the use of boxing.

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/expression/predicate/In.java:
```diff
@@ -61,22 +72,75 @@ public DataType dataType() {
     @Override
     public boolean foldable() {
-        return foldable;
+        return children().stream().allMatch(Expression::foldable);
```
Can be moved to `Expressions#foldable` similar to `nullable` and `resolvable`.

matriv commented:
@costin Addressed comments. Please take another look.

If the whole `IN` expression is not foldable then the values in the list `IN(v1, v2, ...)` are not folded because the `ConstantFolding` is a `DOWN` rule. This is not a problem as when we build the `TermsQuery` or the painless script we call `fold()`.

Making the `ConstantFolding` an `UP` rule solves this more elegantly, but it generates a serious of other issues. Maybe we can introduce a special ConstantFolding rule just for the `IN` or you think the current solution is ok?

costin commented on x-pack/plugin/sql/src/main/java/org/elasticsearch/xpack/sql/querydsl/query/TermsQuery.java:
```diff
@@ -0,0 +1,59 @@
+    public TermsQuery(Location location, String term, List<Expression> values) {
+        super(location);
+        this.term = term;
+        this.values = values.stream().map(Expression::fold).collect(Collectors.toCollection(LinkedHashSet::new));
```
As an alternative, use `Foldables.*`: `new LinkedHashSet(Foldables.valuesOf(values, datatType()))`

matriv commented:
retest this please

matriv commented:
@costin as discussed, I was wrong the optimisation kicks in, it was just not tested properly, I now have a test for that. Thank you!

matriv commented:
retest this please

matriv commented:
retest this please

matriv commented:
Backported to `6.x` with 528e2b27d17df43a8c4e267cbda8e9a673e50343
